Fold mul (lshr exact (X, N)), 2^N + 1 to add (X , lshr exact (X, N))

AZero13 opened the pull request:
Alive2 Proofs:
https://alive2.llvm.org/ce/z/aJnxyp
https://alive2.llvm.org/ce/z/dyeGEv

AZero13 committed:
[InstCombine] Pre-commit tests (NFC)

AtariDreams commented:
@dtcxzyw Thoughts on this?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,43 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
   }
 
+  {
+    // mul (lshr exact (X, 2^N + 1)), N -> add (X , lshr (X, N))
```
```suggestion
    // mul (lshr exact X, N), (2^N + 1) -> add (X , lshr (X, N))
```

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,43 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
+                AddOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "", true);
+          } else {
+            BinOp = Builder.CreateAShr(
+                AddOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "", true);
```
comment the `true`.

goldsteinn commented on llvm/test/Transforms/InstCombine/ashr-lshr.ll:
```diff
@@ -862,4 +862,130 @@ define i32 @ashr_mul_times_5_div_4_exact_2(i32 %x) {
+  %a = ashr exact i32 %x, 3
+  %res = mul nsw i32 %a, 9
+  ret i32 %res
+}
```
Is this case correct?
https://alive2.llvm.org/ce/z/Vag8Wh

AtariDreams replied:
It's not. I'm gonna fix it today.

goldsteinn replied:
I don't think you have solved the problem that you can't propagate `nsw` to the add if the ShiftAmt may be `BitWidth - 1` or `MulAp` may be INT_MIN. This case might get simplified beforehand, but it still shouldn't break w/ your code.

AtariDreams replied:
Fixed!

AtariDreams commented:
Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn 

Then it should be one use no?

AtariDreams commented:
> > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > 
> 
> > Then it should be one use no?
> 
> 
> 
> Im not sure what you are asking.

See:
![image](https://github.com/llvm/llvm-project/assets/83477269/6112c496-c900-4e61-86d7-017eadba0578)

goldsteinn commented:
> > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> > 
> > 
> > > 
> > 
> > 
> > > Then it should be one use no?
> > 
> > 
> > Im not sure what you are asking.
> 
> See:


As a general rule, don't create more instructions. So in that case, req oneuse.

AtariDreams commented:
> > > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Then it should be one use no?
> 
> > > 
> 
> > > 
> 
> > > Im not sure what you are asking.
> 
> > 
> 
> > See:
> 
> 
> 
> 
> 
> As a general rule, don't create more instructions. So in that case, req oneuse.

And what about the freeze? Does that count?

AtariDreams commented:
> > > > Okay so if the IR just repeats I guess I can remove one use but what if there's frozen involved @goldsteinn
> 
> > > 
> 
> > > 
> 
> > > > 
> 
> > > 
> 
> > > 
> 
> > > > Then it should be one use no?
> 
> > > 
> 
> > > 
> 
> > > Im not sure what you are asking.
> 
> > 
> 
> > See:
> 
> 
> 
> 
> 
> As a general rule, don't create more instructions. So in that case, req oneuse.

I'll just reuse Op0 directly I think hopefully it works

goldsteinn commented:
Okay, this LGTM. Please wait on an additional approval to push.

AtariDreams commented:
@nikic Thoughts?

AtariDreams commented:
@nikic Can we please merge this?

AtariDreams commented:
@dtcxzyw I do not have merge permissions.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
+    const APInt *ShiftC;
+    const APInt *MulAP;
+    if ((HasNSW || HasNUW) &&
+        match(&I, m_Mul(m_CombineOr(m_LShr(m_Value(NewOp), m_APInt(ShiftC)),
+                                    m_AShr(m_Value(NewOp), m_APInt(ShiftC))),
```
```suggestion
        match(&I, m_Mul(m_Shr(m_Value(NewOp), m_APInt(ShiftC)),
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
+                                    m_AShr(m_Value(NewOp), m_APInt(ShiftC))),
+                        m_APInt(MulAP)))) {
+      if (BitWidth > 2 && (*MulAP - 1).isPowerOf2() &&
+          MulAP->logBase2() == ShiftC->getZExtValue()) {
```
```suggestion
          ShiftC == MulAP->logBase2()) {
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
+          if (HasNUW && OpBO->getOpcode() == Instruction::AShr &&
+              OpBO->hasOneUse())
+            BinOp = Builder.CreateLShr(
+                NewOp, ConstantInt::get(Ty, ShiftC->getZExtValue()), "",
```
```suggestion
                NewOp, ConstantInt::get(Ty, ShiftC), "",
```

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineMulDivRem.cpp:
```diff
@@ -255,6 +255,38 @@ Instruction *InstCombinerImpl::visitMul(BinaryOperator &I) {
+      if (BitWidth > 2 && (*MulAP - 1).isPowerOf2() &&
+          MulAP->logBase2() == ShiftC->getZExtValue()) {
+        BinaryOperator *OpBO = cast<BinaryOperator>(Op0);
+        if (OpBO->isExact()) {
```
Use `m_Exact` in the pattern match instead.

nikic reviewed (commented):
The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

AtariDreams commented:
The freeze instructions added more instructions so I removed handling for that and now nsw or nuw is required. Freeze is for when none of those are present.

AtariDreams commented:
> The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

Is it okay to use freeze even if that counts as an extra instruction in the IR? Because that's why did a last second replacement with a check for NUW/NSW for the optimization to apply at all.

AtariDreams commented:
> The code doesn't match the proofs (the proofs have freeze instructions, the code does not).

Fixed!

AtariDreams commented:
@dtcxzyw @nikic What do you think about this?

AZero13 committed:
[InstCombine] Fold mul (lshr exact (X, N)), 2^N + 1 -> add (X , lshr exact (X, N))

Alive2 Proofs:
https://alive2.llvm.org/ce/z/aJnxyp
https://alive2.llvm.org/ce/z/dyeGEv

RSilicon commented:
Okay we are back!

RSilicon commented:
> Okay, this LGTM. Please wait on an additional approval to push.

Thoughts?

RSilicon commented:
@dtcxzyw @goldsteinn Is this good now?

DTeachs commented:
@nikic Did you see this yet?

RSilicon commented:
W

RSilicon commented:
@dtcxzyw as I was saying, would you please take a look and approve this please.

RSilicon commented:
Closing for now due to lack of interest from maintainers.

AreaZR commented:
@nikic can we please merge this and be done with it?

AreaZR commented:
@goldsteinn Thoughts?

AreaZR commented:
@goldsteinn @dtcxzyw Thoughts?

AreaZR commented:
@dtcxzyw, could you please review and approve this? Thank you!

AreaZR commented:
@nikic Could you please review this when you get the chance? Thank you.
